src/plots/polar/polar.js

```diff
@@ -1116,26 +1278,229 @@ function computeSectorBBox(sector) {
     return [x0, y0, x1, y1];
 }
 
-function pathSector(r, sector) {
-    if(isFullCircle(sector)) {
-        return Drawing.symbolFuncs[0](r);
+function isAngleInSector(rad, sector) {
+    if(isFullCircle(sector)) return true;
+
+    var s0 = wrap360(sector[0]);
+    var s1 = wrap360(sector[1]);
+    if(s0 > s1) s1 += 360;
+
+    var deg = wrap360(rad2deg(rad));
+    var nextTurnDeg = deg + 360;
+
+    return (deg >= s0 && deg <= s1) ||
+        (nextTurnDeg >= s0 && nextTurnDeg <= s1);
+}
+
+function snapToVertexAngle(a, vangles) {
+    function angleDeltaAbs(va) {
+        return Math.abs(angleDelta(a, va));
     }
 
-    var xs = r * Math.cos(deg2rad(sector[0]));
-    var ys = -r * Math.sin(deg2rad(sector[0]));
-    var xe = r * Math.cos(deg2rad(sector[1]));
-    var ye = -r * Math.sin(deg2rad(sector[1]));
+    var ind = findIndexOfMin(vangles, angleDeltaAbs);
+    return vangles[ind];
+}
 
-    var arc = Math.abs(sector[1] - sector[0]);
-    var flags = arc <= 180 ? [0, 0, 0] : [0, 1, 0];
+// taken from https://stackoverflow.com/a/2007279
+function angleDelta(a, b) {
+    var d = b - a;
+    return Math.atan2(Math.sin(d), Math.cos(d));
+}
+
+function findIndexOfMin(arr, fn) {
+    fn = fn || Lib.identity;
+
+    var min = Infinity;
+    var ind;
+
+    for(var i = 0; i < arr.length; i++) {
+        var v = fn(arr[i]);
+        if(v < min) {
+            min = v;
+            ind = i;
+        }
+    }
+    return ind;
+}
+
+// find intersection of 'v0' <-> 'v1' edge with a 'radial' line
+// (i.e. a line that starts from the origin at angle 'a')
+// given an (xp,yp) pair on the 'v0' <-> 'v1' line
+// (N.B. 'v0' and 'v1' are angles in radians)
```

Comment:
> whew, I think I understand the goal of this function: so you have two rays at angles v0 and v1, and given points on each ray at equal radius and a segment connecting them, first find the radius (or the segment) for which that segment intersects [xp, yp]. Then, find the point on a ray at angle a that intersects this segment. Did I get that right?

Yeah, you got it right except that a priori we don't know the points on each ray. We only know their angles. 

> 
If a always matches either v0 or v1 this looks fine, but the else clause looks like it will give the wrong answer if a is anything else (since the norm of the result won't be r then).

Good eye. Thanks! 

When computing the the zoombox coordinates, `a` matches `v0` so no bugs there. When computing the (clipped) polygon coordinates (when `polar.sector` isn't [0,360]) `a` is either `polar.sector[0]` or `polar.sector[1]`.

Reply:
At lot of little things weren't perfect for non-[0,360] polar sectors. Thanks for bringing this up. https://github.com/plotly/plotly.js/pull/2739/commits/7d405a04a7298f74ef2533b2951b2a1dd72fb3ab?short_path=c565e02#diff-c565e0252037c7b20b8005938d85799a fixes this.